server/task/src/main/java/org/apache/james/task/MemoryTaskManagerWorker.java

```diff
@@ -0,0 +1,114 @@
+ * software distributed under the License is distributed on an  *
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY       *
+ * KIND, either express or implied.  See the License for the    *
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ ****************************************************************/
+package org.apache.james.task;
+
+import java.util.Optional;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.function.BiConsumer;
+import java.util.function.Consumer;
+
+import org.apache.james.util.MDCBuilder;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import reactor.core.publisher.Mono;
+
+public class MemoryTaskManagerWorker implements TaskManagerWorker {
+    private static final boolean INTERRUPT_IF_RUNNING = true;
+    private static final Logger LOGGER = LoggerFactory.getLogger(MemoryTaskManagerWorker.class);
+    private final ConcurrentHashMap<TaskId, CompletableFuture<Task.Result>> idToFuture = new ConcurrentHashMap<>();
+
+    @Override
+    public Mono<Task.Result> executeTask(TaskWithId taskWithId, Consumer<TaskExecutionDetailsUpdater> updateDetails) {
+        CompletableFuture<Task.Result> futureResult = CompletableFuture.supplyAsync(() -> runWithMdc(taskWithId, updateDetails));
+
+        idToFuture.put(taskWithId.getId(), futureResult);
+
+        Mono<Task.Result> result = Mono.<Task.Result>fromFuture(futureResult)
+            .doOnError(res -> failed(updateDetails,
+                (logger, details) -> logger.error("Task was partially performed. Check logs for more details")))
+            .doOnTerminate(() -> idToFuture.remove(taskWithId.getId()));
+
+        return result;
+    }
+
+    private Task.Result runWithMdc(TaskWithId taskWithId, Consumer<TaskExecutionDetailsUpdater> updateDetails) {
+        return MDCBuilder.withMdc(
+            MDCBuilder.create()
+                .addContext(Task.TASK_ID, taskWithId.getId())
+                .addContext(Task.TASK_TYPE, taskWithId.getTask().type())
+                .addContext(Task.TASK_DETAILS, taskWithId.getTask().details()),
+            () -> run(taskWithId, updateDetails));
+    }
+
+
+    private Task.Result run(TaskWithId taskWithId, Consumer<TaskExecutionDetailsUpdater> updateDetails) {
+        updateDetails.accept(TaskExecutionDetails::start);
+        try {
+            return taskWithId.getTask()
+                .run()
+                .onComplete(() -> success(updateDetails))
+                .onFailure(() -> failed(updateDetails, (logger, details) -> logger.error("Task was partially performed. Check logs for more details" + details.getTaskId())));
+        } catch (Exception e) {
+            failed(updateDetails,
+                (logger, executionDetails) -> logger.error("Error while running task", executionDetails, e));
```

Comment:
s/"Error while running task"/"Error while running task {}"

Reply:
then use executionDetails.toString() ?

Reply:
I didn't change that. Just moved from the manager to the worker.
It seems the toString method of TaskExecutionDetails is not overriden too.